posthog/clickhouse/query_fragment.py

```diff
@@ -0,0 +1,99 @@
+import itertools
+import re
+from dataclasses import dataclass
+from functools import cached_property
+from typing import Any, Dict, List, Mapping, Sequence, Tuple, Union
+
+ParamValueType = Union[str, int, float, None, List]
+QueryFragmentLike = Union["QueryFragment", str]
+
+unique_sequence = itertools.count()
+
+
+@dataclass(frozen=True)
+class Param:
+    value: ParamValueType
+
+
+@dataclass
+class QueryFragment:
+    sql: str
+    params: Dict[str, Param]
+
+    def __init__(
+        self,
+        sql: Union[str, "QueryFragment"],
+        params_and_fragments: Mapping[str, Union["Param", "QueryFragment"]] = {},
+        **kwargs: Union["Param", "QueryFragment"],
+    ):
+        if len(kwargs) > 0 and len(params_and_fragments) > 0:
```

Comment:
Not sold on this either. Idea was that kwargs are easier/prettier to write than dicts in some cases but you can't do `UniqueName`s with kwargs at least with the current pattern.

Reply:
Proposal: kwargs for everything, and a separate parameter for unique names only? These can be merged together freely because uniqueNames values should be params only.